File: src/luademo/lua_demo_actor.ts
import { assign_bt, insavegame, WorldObject, type RevivableObjectArgs } from 'bmsx';
import { clamp } from 'bmsx/utils/utils';

type BehaviorSummary = {
	mode: string;
	status: string;
	pulse: number;
	iteration: number;
	hue: number;
	interval: number;
};

@insavegame
@assign_bt('lua_demo_bt')
export class LuaDemoActor extends WorldObject {
	public behavior: BehaviorSummary;

	constructor(opts?: RevivableObjectArgs & { id?: string }) {
		const initOpts = opts ? { ...opts, id: opts.id ?? 'lua_demo_actor' } : { id: 'lua_demo_actor' };
		super(initOpts);
		this.behavior = {
			mode: 'boot',
			status: 'Priming behavior tree...',
			pulse: 0,
			iteration: 0,
			hue: 9,
			interval: 0,
		};
	}

	public resetBehavior(): void {
		this.behavior.mode = 'boot';
		this.behavior.status = 'Priming behavior tree...';
		this.behavior.pulse = 0;
		this.behavior.iteration = 0;
		this.behavior.hue = 9;
		this.behavior.interval = 0;
	}

	public setMode(mode: string): void {
		this.behavior.mode = mode;
	}

	public setBehaviorStatus(status: string): void {
		this.behavior.status = status;
	}

	public adjustPulse(delta: number): number {
		this.behavior.pulse = clamp(this.behavior.pulse + delta, 0, 1);
		return this.behavior.pulse;
	}

	public setPulse(value: number): number {
		this.behavior.pulse = clamp(value, 0, 1);
		return this.behavior.pulse;
	}

	public setHue(hue: number): void {
		const quantized = Math.floor(hue);
		this.behavior.hue = clamp(quantized, 1, 15);
	}

	public incrementIteration(): number {
		this.behavior.iteration += 1;
		return this.behavior.iteration;
	}

	public setCurrentInterval(frames: number): void {
		const quantized = Math.floor(frames);
		this.behavior.interval = clamp(quantized, 0, Number.MAX_SAFE_INTEGER);
	}
}

(globalThis as Record<string, unknown>).LuaDemoActor = LuaDemoActor;

File: src/carts/luademo/luademo_actor.lua
local LuaDemoActor = LuaDemoActor or {}

local function clamp(value, min_value, max_value)
	if value < min_value then
		return min_value
	end
	if value > max_value then
		return max_value
	end
	return value
end

local function ensure_behavior(target)
	local behavior = target.behavior
	if not behavior then
		behavior = {}
		target.behavior = behavior
	end
	if behavior.mode == nil then
		behavior.mode = 'idle'
	end
	if behavior.hue == nil then
		behavior.hue = 10
	end
	if behavior.pulse == nil then
		behavior.pulse = 0
	end
	if behavior.iteration == nil then
		behavior.iteration = 0
	end
	if behavior.interval == nil then
		behavior.interval = 0
	end
	if behavior.status == nil then
		behavior.status = ''
	end
	return behavior
end

function LuaDemoActor:create(actor)
	actor.visible = true
	actor.behavior = nil
	LuaDemoActor.resetbehavior(actor)
end

function LuaDemoActor:resetbehavior()
	local behavior = ensure_behavior(self)
	behavior.mode = 'priming'
	behavior.hue = 10
	behavior.pulse = 0
	behavior.iteration = 0
	behavior.interval = 0
	behavior.status = 'Booting behavior tree'
	return behavior
end

function LuaDemoActor:setmode(mode)
	local behavior = ensure_behavior(self)
	behavior.mode = mode or 'idle'
end

function LuaDemoActor:sethue(hue)
	local behavior = ensure_behavior(self)
	local next_hue = hue or 0
	behavior.hue = clamp(math.floor(next_hue + 0.5), 0, 15)
end

function LuaDemoActor:setbehaviorstatus(text)
	local behavior = ensure_behavior(self)
	if text == nil then
		behavior.status = ''
	else
		behavior.status = tostring(text)
	end
end

function LuaDemoActor:adjustpulse(delta)
	local behavior = ensure_behavior(self)
	local amount = delta or 0
	local next_value = clamp((behavior.pulse or 0) + amount, 0, 1)
	behavior.pulse = next_value
	return next_value
end

function LuaDemoActor:setcurrentinterval(interval)
	local behavior = ensure_behavior(self)
	local next_interval = interval or 0
	if next_interval < 0 then
		next_interval = 0
	end
	behavior.interval = math.floor(next_interval + 0.5)
end

function LuaDemoActor:incrementiteration()
	local behavior = ensure_behavior(self)
	local next_value = (behavior.iteration or 0) + 1
	behavior.iteration = next_value
	return next_value
end

register_worldobject({
	id = 'LuaDemoActor',
	class = LuaDemoActor,
	bts = {
		{ id = 'lua_demo_bt', auto_tick = true },
	},
})
